### PR TITLE
Add PytorchBatchConverter

### PR DESCRIPTION
**Pull request by yeandy**
Adding `PytorchBatchConverter` as [discussed](https://github.com/apache/beam/issues/21440#issuecomment-1239777760) in https://github.com/apache/beam/issues/21440 as an initial step to integrate Batched DoFns into RunInference.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/python/apache_beam/typehints/batch_test.py [hunk 1]
```diff
@@ -54,6 +64,17 @@
         'element_typehint': str,
         'batch': ["foo" * (i % 5) + str(i) for i in range(1000)],
     },
+    {
+        'batch_typehint': torch.Tensor,
+        'element_typehint': torch.Tensor,
```
Comment (yeandy) [inline sdks/python/apache_beam/typehints/batch_test.py:69]: A single element of a `Tensor` array is also a `Tensor`. e.g.
```
>>> a = torch.Tensor([1,2,3])
>>> a
tensor([1., 2., 3.])
>>> a[0]
tensor(1.)
>>> type(a)
<class 'torch.Tensor'>
>>> type(a[0])
<class 'torch.Tensor'>
```

**Comment by yeandy** [comment]
R: @TheNeuralBit

File: sdks/python/apache_beam/typehints/batch_test.py (hunk 1 above)
Comment (TheNeuralBit) [reply sdks/python/apache_beam/typehints/batch_test.py:69]: Interesting. It could be problematic to allow this as an element type though, since it's unclear what data type is.

For now, could we always represent scalars as a 0-dim PytorchTensor? i.e. `PytorchTensor[torch.int32, (,)]`. There could also be a wrapper for this, like `PytorchScalar[torch.int32]`

File: sdks/python/apache_beam/typehints/batch.py [hunk 2]
```diff
@@ -35,6 +35,7 @@
 from typing import TypeVar
 
 import numpy as np
+import torch
```
Comment (TheNeuralBit) [inline sdks/python/apache_beam/typehints/batch.py:38]: Could you make this a separate module, following the pattern I just used for pandas: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/typehints/pandas_type_compatibility.py

That way it can be separable from everything else here if `torch` is available.

Comment (TheNeuralBit) [reply sdks/python/apache_beam/typehints/batch.py:38]: One issue with that is I haven't found a good way to re-use the BatchConverterTest logic, for now it's just duplicated in `pandas_type_compatibility_test`.

**Review by TheNeuralBit — COMMENTED** [review]
Thanks @yeandy, it's great to see this! I have a couple minor requests.

File: sdks/python/apache_beam/typehints/batch_test.py (hunk 1 above)
Comment (yeandy) [reply sdks/python/apache_beam/typehints/batch_test.py:69]: Yeah, let me try that

File: sdks/python/apache_beam/typehints/batch.py (hunk 2 above)
Comment (yeandy) [reply sdks/python/apache_beam/typehints/batch.py]: Done.

**Comment by yeandy** [comment]
PTAL @TheNeuralBit

**Review by TheNeuralBit — APPROVED** [review]
LGTM, thank you @yeandy

File: sdks/python/apache_beam/typehints/pytorch_type_compatibility_test.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    },
+    {
+        'batch_typehint': PytorchTensor[torch.int64, (N, 10)],
+        'element_typehint': PytorchTensor[torch.int64, (10, )],
+        'batch': torch.tensor([list(range(i, i + 10)) for i in range(100)],
+                              dtype=torch.int64),
+    },
+])
+@pytest.mark.uses_pytorch
+class DataFrameBatchConverterTest(unittest.TestCase):
```
Comment (TheNeuralBit) [inline sdks/python/apache_beam/typehints/pytorch_type_compatibility_test.py:53]: nit: copy-paste error
```suggestion
class PytorchBatchConverterTest(unittest.TestCase):
```

**Comment by TheNeuralBit** [comment]
It looks like codecov is unhappy since this file looks untested. Presumably because we don't have pytorch installed in the run generating coverage data. I confirmed this test is exercised on Jenkins: 
image
